### PR TITLE
Update documentation of hooks LIFO order 

### PR DESCRIPTION
**Pull request by noklam**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

Summarising https://github.com/orgs/kedro-org/projects/2/views/4?filterQuery=-no%3Astatus+sprint%3A%22%40current%22+assignee%3Anoklam+&pane=issue&itemId=12463585, the scope of this ticket is limited to the LIFO order, #1718 will be the ticket to summarise the order of `kedro run` 

(Opening this PR to draft my idea and I will add the docs accordingly and update the title)
Few things I plan to do:
-  [x] We need to also document what the order is in which auto-discovered hooks run. Add this to: https://kedro.readthedocs.io/en/stable/hooks/introduction.html To verify the run order, create a project and install several plugins with hooks to test.
-  [x] Add a comment in the settings.py template file to explain the run order of hooks 
- Check if https://github.com/pytest-dev/pytest/blob/196f01965ed264e1fd917db236c872f85403753d/doc/en/how-to/writing_plugins.rst#plugin-discovery-order-at-tool-startup helps or link to pytest for more detail
-  [x]  Maybe this comment https://github.com/kedro-org/kedro/issues/1718#issuecomment-1551643867
- [x] https://docs.kedro.org/en/stable/extend_kedro/plugins.html#hooks - I think we missed to update the documentation when we move from `pyproject.toml` to `setup.py`. I want to keep both in the documentation since people aren't necessary moving to `pyproject.toml` (existing projects). I found this https://setuptools.pypa.io/en/latest/userguide/entry_point.html#entry-points-for-plugins
-  [ ] Do the same for starters (After this Pr is approved) - https://github.com/kedro-org/kedro-starters/pull/146
- [ ] Maybe update `_register_hooks_setuptools`, it's outdated.
## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Comment by noklam** [comment]
https://docs.kedro.org/en/stable/extend_kedro/plugins.html#community-developed-plugins And while I am looking at this, we have `awesome-kedro` now, maybe it make sense to shuffle the order accordingly or at least move the more stable one to the top? (kedro-mlflow and GetInDatas)

cc @stichbury

**Comment by stichbury** [comment]
> https://docs.kedro.org/en/stable/extend_kedro/plugins.html#community-developed-plugins And while I am looking at this, we have `awesome-kedro` now, maybe it make sense to shuffle the order accordingly or at least move the more stable one to the top? (kedro-mlflow and GetInDatas)
> 
> cc @stichbury

That sounds OK to me, although it's hard to rank them by stability once we get past the most obvious. We did have a discussion a while ago that @idanov initiated about using a badge on individual plugin README pages to indicate when they were last updated. But anyway, let's remove the list from the docs and point instead to `awesome-kedro` where we list the plugins with GetInData starting the list and a rough order once we've included the most obviously stable.

**Comment by noklam** [comment]
After some testing, I found out the LIFO order is quite tricky,  and I don't recommend we document it as is.

The "registration order" is a combination of these factors:
1. Project hooks
2. hooks from entrypoints (which follow alphabetical order, and is a 2-level for loop

The execution order is the reverse of registration order (LIFO), except that user may use `hook_impl` "try_first" argument. Optionally, you can also import a registered entrypoints into project to change the order (i.e. importing kedro_viz hook in `settings.py` will force it to be registered earlier)

In general, order shouldn't be matter (99% of the time). If user need to control this, we should recommend using `pluggy` built-in functionality "try_first" argument. We shouldn't recommend to control the order via `entrypoints` order

File: docs/source/extend_kedro/plugins.md [hunk 1]
```diff
@@ -29,8 +29,15 @@ def to_json(metadata):
     pipeline = pipelines["__default__"]
     print(pipeline.to_json())
 ```
+Starting from 0.18.14, Kedro switch to replace `setup.py` with `pyproject.toml`. The plugin need to provide the entry points in either file.
```
Comment (stichbury) [inline docs/source/extend_kedro/plugins.md:32]:
```suggestion
From version 0.18.14, Kedro replaced `setup.py` with `pyproject.toml`. The plugin needs to provide entry points in either file.
```

File: docs/source/extend_kedro/plugins.md [hunk 2]
```diff
@@ -29,8 +29,15 @@ def to_json(metadata):
...
 
-The plugin provides the following `entry_points` config in `setup.py`:
+To use `pyproject.toml`, specifiy
```
Comment (stichbury) [inline docs/source/extend_kedro/plugins.md:34]:
```suggestion
To add the entry point to `pyproject.toml`, specify it as follows:
```

File: docs/source/extend_kedro/plugins.md [hunk 3]
```diff
@@ -29,8 +29,15 @@ def to_json(metadata):
...
+```toml
+[project.entry-points."kedro.project_commands"]
+kedrojson = kedrojson.plugin.commands
+```
+
+To use `setup.py`, the plugin need to provide the following `entry_points` config in `setup.py`:
```
Comment (stichbury) [inline docs/source/extend_kedro/plugins.md:40]:
```suggestion
To use `setup.py`, the plugin needs to provide the following `entry_points` configuration in `setup.py`:
```

File: docs/source/extend_kedro/plugins.md [hunk 4]
```diff
@@ -162,6 +186,12 @@ You can also develop hook implementations to extend Kedro's CLI behaviour in you
 setup(entry_points={"kedro.cli_hooks": ["plugin_name = plugin_name.plugin:cli_hooks"]})
 ```
 
+Alternatively, to use `pyproject.toml`
```
Comment (stichbury) [inline docs/source/extend_kedro/plugins.md:189]:
```suggestion
Alternatively, to use `pyproject.toml`:
```

File: docs/source/extend_kedro/plugins.md [hunk 5]
```diff
@@ -162,6 +186,12 @@ You can also develop hook implementations to extend Kedro's CLI behaviour in you
...
+```toml
+[project.entry-points."kedro.cli_hooks"]
+plugin_name = plugin_name.plugin.cli_hooks
+```
+
 where `plugin.py` is the module where you declare hook implementations:
```
Comment (stichbury) [inline docs/source/extend_kedro/plugins.md:195]:
```suggestion
(where `plugin.py` is the module where you declare hook implementations):
```

File: docs/source/extend_kedro/plugins.md [hunk 6]
```diff
@@ -204,28 +234,17 @@ connectors are implementations of the `AbstractDataset`
 
 ## Community-developed plugins
 
-See the full list of plugins using the GitHub tag [kedro-plugin](https://github.com/topics/kedro-plugin).
+See the full list of plugins using the GitHub repository [awesome-kedro](https://github.com/kedro-org/awesome-kedro?tab=readme-ov-file#kedro-plugins).
```
Comment (stichbury) [inline docs/source/extend_kedro/plugins.md:237]:
```suggestion
A comprehensive list of plugins is published on the [`awesome-kedro`](https://github.com/kedro-org/awesome-kedro?tab=readme-ov-file#kedro-plugins) GitHub repository.
```

**Review by stichbury — CHANGES_REQUESTED** [review]
Few changes suggested and I'll make a commit direct onto `hooks.md` on your branch if that's OK as I want to make a few changes there.

**Comment by stichbury** [comment]
@ankatiyar I know I've asked this before but @noklam has touched a few pages of markdown and there were some issues with his text but Vale didn't flag anything.  I'm going to deliberately introduce an issue into `hooks.md` to see if it shouts...I'm not convinced it's running as it should be.

**Comment by ankatiyar** [comment]
> @ankatiyar I know I've asked this before but @noklam has touched a few pages of markdown and there were some issues with his text but Vale didn't flag anything. I'm going to deliberately introduce an issue into `hooks.md` to see if it shouts...I'm not convinced it's running as it should be.

I recently changed the settings on the action to only run when a PR is opened or reopened because the vale linter was leaving a whole set of review comments (even the ones that were dismissed) with every push or merging of main to branch which was getting very hard to deal with.
I can add a manual trigger to the linter so you can trigger it for a specific branch.

File: docs/source/hooks/introduction.md [hunk 7]
```diff
@@ -126,6 +132,19 @@ DISABLE_HOOKS_FOR_PLUGINS = ("<plugin_name>",)
...
+
+In general, Hook execution order is not guaranteed and you should not rely on it. If you need to make sure a particular Hook is executed first or last, you can use the the [`tryfirst` or `trylast` argument](https://pluggy.readthedocs.io/en/stable/index.html#call-time-order) for `hook_impl`.
+
 ## Under the hood
 
-Under the hood, we use [pytest's pluggy](https://pluggy.readthedocs.io/en/latest/) to implement Kedro's Hook mechanism. We recommend reading their documentation if you have more questions about the underlying implementation.
+Under the hood, we use [pytest's pluggy](https://pluggy.readthedocs.io/en/latest/) to implement Kedro's Hook mechanism. We recommend reading their documentation to find out more about the underlying implementation.
+
+### Plugin Hooks
+Plugin Hooks are registered using []`importlib_metadata`'s `EntryPoints` API](https://docs.python.org/3/library/importlib.metadata.html).
```
Comment (merelcht) [inline docs/source/hooks/introduction.md:150]:
```suggestion
Plugin Hooks are registered using [`importlib_metadata`'s `EntryPoints` API](https://docs.python.org/3/library/importlib.metadata.html).
```

File: docs/source/hooks/introduction.md [hunk 8]
```diff
@@ -126,6 +132,19 @@ DISABLE_HOOKS_FOR_PLUGINS = ("<plugin_name>",)
 
 where `<plugin_name>` is the name of an installed plugin for which the auto-registered Hooks must be disabled.
 
+## Hook execution order
+Hooks follow a Last-In-First-Out (LIFO) order, which means the first registered Hook will be executed last.
+
+Hooks are registered in the following order:
+
+1. Project Hooks in `settings.py` - If you have `HOOKS = (hook_a, hookb,)`, `hook_b` will be executed before `hook_a`
```
Comment (merelcht) [inline docs/source/hooks/introduction.md:140]:
```suggestion
1. Project Hooks in `settings.py` - If you have `HOOKS = (hook_a, hook_b,)`, `hook_b` will be executed before `hook_a`
```

File: docs/source/extend_kedro/plugins.md (hunk 1 above)
Comment (noklam) [reply docs/source/extend_kedro/plugins.md:32]: @stichbury We have moved to `pyproject.toml` if you created starter with the new version, but there are projects created with older version of Kedro and they only upgrade `kedro` but not the project template. Maybe its's too confusing and I should just keep the new one?

**Comment by noklam** [comment]
I have addressed the comment.

> I think the problem is that it is commenting once but then if someone adds a new commit it dismisses all the comments so they never show up by the time I get there, and it doesn't re-run on new additions. It does need to keep running on each commit to check new changes. What we don't want is for it to re-publish dismissed comments, but anything not addressed needs to be retained.

I agree this is ideal but probably need some more effort to setup.  Meanwhile I add a `make language-lint` for triggering manually, optionally support to run at specific directory since there are many warnings not related to the PR if I run at `docs/`.

**Comment by noklam** [comment]
As far as I understand, the `pluggy` PluginManager work with entrypoints beyond `setuptools`. `PluginManager.load_setuptools_entrypoints` is misleading. Internally, we have functions likes `load_entry_points` for CLI.
@astrojuanlu

File: kedro/framework/hooks/manager.py [hunk 9]
```diff
@@ -73,6 +73,7 @@ def _register_hooks_setuptools(
     plugin_names = set()
     disabled_plugin_names = set()
     for plugin, dist in plugininfo:
+        print("DEBUG!!!", plugin, dist)
```
Comment (merelcht) [inline kedro/framework/hooks/manager.py:76]: I'm guessing you only added this for debugging? 😄

File: RELEASE.md [hunk 10]
```diff
@@ -16,6 +16,7 @@
 * Updated dataset factories to resolve nested catalog config properly.
 
 ## Documentation changes
+* Added documentation to clarify execution orders of hooks.
```
Comment (merelcht) [inline RELEASE.md:19]:
```suggestion
* Added documentation to clarify execution order of hooks.
```

**Review by merelcht — APPROVED** [review]
Thanks for also updating the docs about plugins!

File: docs/source/extend_kedro/plugins.md [hunk 11]
```diff
@@ -127,10 +125,14 @@ We use the following command convention: `kedro <plugin-name> <command>`, with `
 
 ## Hooks
 
-You can develop hook implementations and have them automatically registered to the project context when the plugin is installed. To enable this for your custom plugin, simply add the following entry in your `setup.py`:
+You can develop hook implementations and have them automatically registered to the project context when the plugin is installed.
 
-```python
-setup(entry_points={"kedro.hooks": ["plugin_name = plugin_name.plugin:hooks"]})
+To enable this for your custom plugin, simply add the following entry in your `pyproject.toml`
```
Comment (stichbury) [inline docs/source/extend_kedro/plugins.md:130]:
```suggestion
To enable this for your custom plugin, simply add the following entry in `pyproject.toml`
```

File: docs/source/extend_kedro/plugins.md [hunk 12]
```diff
@@ -127,10 +125,14 @@ We use the following command convention: `kedro <plugin-name> <command>`, with `
...
+
+To use `pyproject.toml`, specifiy
```
Comment (stichbury) [inline docs/source/extend_kedro/plugins.md:132]:
```suggestion
To use `pyproject.toml`, specify
```

File: docs/source/extend_kedro/plugins.md [hunk 13]
```diff
@@ -156,13 +158,15 @@ hooks = MyHooks()
 
 ## CLI Hooks
 
-You can also develop hook implementations to extend Kedro's CLI behaviour in your plugin. To find available CLI hooks, please visit [kedro.framework.cli.hooks](/kedro.framework.cli.hooks). To register CLI hooks developed in your plugin with Kedro, add the following entry in your project's `setup.py`:
+You can also develop hook implementations to extend Kedro's CLI behaviour in your plugin. To find available CLI hooks, please visit [kedro.framework.cli.hooks](/kedro.framework.cli.hooks). To register CLI hooks developed in your plugin with Kedro, add the following entry in your project's `pyproject.toml`:
```
Comment (stichbury) [inline docs/source/extend_kedro/plugins.md:161]:
```suggestion
You can also develop Hook implementations to extend Kedro's CLI behaviour in your plugin. To find available CLI Hooks, please visit [kedro.framework.cli.hooks](/kedro.framework.cli.hooks). To register CLI Hooks developed in your plugin with Kedro, add the following entry in your project's `pyproject.toml`:
```

File: docs/source/extend_kedro/plugins.md [hunk 14]
```diff
@@ -156,13 +158,15 @@ hooks = MyHooks()
...
-```python
-setup(entry_points={"kedro.cli_hooks": ["plugin_name = plugin_name.plugin:cli_hooks"]})
+
+```toml
+[project.entry-points."kedro.cli_hooks"]
+plugin_name = plugin_name.plugin.cli_hooks
 ```
 
-where `plugin.py` is the module where you declare hook implementations:
+(where `plugin.py` is the module where you declare hook implementations):
```
Comment (stichbury) [inline docs/source/extend_kedro/plugins.md:169]:
```suggestion
(where `plugin.py` is the module where you declare Hook implementations):
```

File: docs/source/hooks/examples.md [hunk 15]
```diff
@@ -298,7 +298,7 @@ class PipelineMonitoringHooks:
         self._client.incr("run")
 ```
 
-* Register Hooks implementation, as described in the [hooks documentation](introduction.md#registering-your-hook-implementations-with-kedro) and run Kedro.
+* Register Hooks implementation, as described in the [hooks documentation](introduction.md#registering-the-hook-implementation-with-kedro) and run Kedro.
```
Comment (stichbury) [inline docs/source/hooks/examples.md:301]:
```suggestion
* Register the Hooks implementation, as described in the [Hooks documentation](./introduction.md#registering-the-hook-implementation-with-kedro) and run Kedro.
```

File: docs/source/hooks/examples.md [hunk 16]
```diff
@@ -365,7 +365,7 @@ class ModelTrackingHooks:
         mlflow.end_run()
 ```
 
-* Register Hooks implementation, as described in the [hooks documentation](introduction.md#registering-your-hook-implementations-with-kedro) and run Kedro.
+* Register Hooks implementation, as described in the [hooks documentation](introduction.md#registering-the-hook-implementation-with-kedro) and run Kedro.
```
Comment (stichbury) [inline docs/source/hooks/examples.md:368]:
```suggestion
* Register the Hook implementation, as described in the [Hooks documentation](./introduction.md#registering-the-hook-implementation-with-kedro) and run Kedro.
```

File: docs/source/hooks/examples.md [hunk 17]
```diff
@@ -409,4 +409,4 @@ In the example above, the `before_node_run` hook implementation must return data
 ```
 
 
-To apply the changes once you have implemented a new hook, you must register it, as described in the [hooks documentation](introduction.md#registering-your-hook-implementations-with-kedro), and then run Kedro.
+To apply the changes once you have implemented a new hook, you must register it, as described in the [hooks documentation](introduction.md#registering-the-hook-implementation-with-kedro), and then run Kedro.
```
Comment (stichbury) [inline docs/source/hooks/examples.md:412]:
```suggestion
Once you have implemented a new Hook you must register it as described in the [Hooks documentation](./introduction.md#registering-the-hook-implementation-with-kedro), and then run Kedro.
```

File: docs/source/hooks/index.md [hunk 18]
```diff
@@ -1,6 +1,6 @@
 # Hooks
 
-Hooks are a mechanism to add extra behaviour to Kedro's main execution in an easy and consistent manner. Some examples might include:
+Hooks are a mechanism to add extra behaviour to Kedro's main execution in a very easy and consistent manner. Some examples might include:
```
Comment (stichbury) [inline docs/source/hooks/index.md:3]:
```suggestion
Hooks are a mechanism to add extra behaviour to Kedro's main execution in an easy and consistent manner. Some examples might include:
```

File: docs/source/hooks/introduction.md [hunk 19]
```diff
@@ -36,18 +33,24 @@ The naming convention for error hooks is `on_<noun>_error`, in which:
 [kedro.framework.hooks](/kedro.framework.hooks) lists the full specifications for which you can inject additional behaviours by providing an implementation.
 
 
-#### CLI hooks
+### CLI hooks
```
Comment (stichbury) [inline docs/source/hooks/introduction.md:36]:
```suggestion
### CLI Hooks
```

**Review by stichbury — APPROVED** [review]
LGTM, just a few changes which are mostly capitalisation of Hook 😆

File: kedro/framework/hooks/manager.py (hunk 9 above)
Comment (noklam) [reply kedro/framework/hooks/manager.py:76]: Thanks! 😅 I was working between different branches must have left this accidentally.

File: docs/source/extend_kedro/plugins.md [hunk 20]
```diff
@@ -29,13 +29,12 @@ def to_json(metadata):
...
-The plugin provides the following `entry_points` config in `setup.py`:
-
-```python
-setup(
-    entry_points={"kedro.project_commands": ["kedrojson = kedrojson.plugin:commands"]}
-)
+To add the entry point to `pyproject.toml`, the plugin needs to provide the following `entry_points` configuration:
+```toml
+[project.entry-points."kedro.project_commands"]
+kedrojson = kedrojson.plugin.commands
```
Comment (astrojuanlu) [inline docs/source/extend_kedro/plugins.md:37]: oops, typo

File: docs/source/extend_kedro/plugins.md [hunk 21]
```diff
@@ -73,12 +72,11 @@ starters = [
...
-In your `setup.py`, you need to register the specifications to `kedro.starters`.
+In your `pyproject.toml`, you need to register the specifications to `kedro.starters`:
 
-```python
-setup(
-    entry_points={"kedro.starters": ["starter =  plugin:starters"]},
-)
+```toml
+[project.entry-points."kedro.starters"]
+starter = plugin.starters
```
Comment (astrojuanlu) [inline docs/source/extend_kedro/plugins.md:79]: Shouldn't this be `plugin:starters`? (Same typo as above)

File: docs/source/extend_kedro/plugins.md (hunk 20 above)
Comment (noklam) [reply docs/source/extend_kedro/plugins.md:37]: I did test it with the plugin hooks🥲 maybe I mess it up when I  copy-paste it back to the doc. Thanks for spotting this.

**Pull request merged**